src/library/scala/collection/immutable/Queue.scala

```diff
@@ -148,7 +153,7 @@ sealed class Queue[+A] protected(protected val in: List[A], protected val out: L
     *  @return the first element of the queue.
     */
   def dequeue: (A, Queue[A]) = out match {
-    case Nil if !in.isEmpty => val rev = in.reverse ; (rev.head, new Queue(Nil, rev.tail))
+    case x :: Nil if in.nonEmpty => (x, new Queue(Nil, in.reverse))
```

Comment:
I'm not sure incurring the reverse eagerly is worth the goal of supporting dubious head/tail pattern described in the ticket. As an aside, there is ongoing conversation about whether to always prefer `!isEmpty` to `nonEmpty` for performance reasons (because cost is not zero).

Reply:
`Queue` is a `LinearSeq` so there is an expectation that `head`/`tail` should have decent performance.

I didn't look at the generated code but going by the `List` implementation I would definitely prefer to call `isEmpty` when the type is known (like it is in this case). The Scala optimizer should be able to inline the method so you don't even have to rely on HotSpot.